src/uct/ib/ud/base/ud_ep.c

```diff
@@ -140,15 +141,97 @@ static ucs_status_t uct_ud_ep_free_by_timeout(uct_ud_ep_t *ep,
+        ucs_assert(!(skb->flags & UCT_UD_SEND_SKB_FLAG_INVALID));
+        if (ucs_likely(!(skb->flags & (UCT_UD_SEND_SKB_FLAG_RESENDING |
+                                       UCT_UD_SEND_SKB_FLAG_COMP)))) {
+            /* fast path case: completed skb without completion callback */
+            uct_ud_skb_release(skb, 1);
+        } else if (ucs_likely(!(skb->flags & UCT_UD_SEND_SKB_FLAG_RESENDING))) {
+            /* completed skb with user completion */
+            ucs_assert(skb->flags & UCT_UD_SEND_SKB_FLAG_COMP);
+            if (ucs_likely(!is_async)) {
+                /* dispatch user completion immediately */
+                uct_ud_iface_dispatch_comp(iface, uct_ud_comp_desc(skb)->comp,
+                                           status);
+                uct_ud_skb_release(skb, 1);
+            } else {
+                /* Don't call user completion from async context. Instead, put
+                 * it on a queue which will be progressed from main thread.
+                 */
+                uct_ud_iface_add_async_comp(iface, skb, status);
+            }
+        } else {
+            /* slow-path case: the skb is still used by the QP, can happen in
+             * when this skb is being resent */
+            ucs_assert(ep->tx.resend_count > 0);
+            skb->flags |= UCT_UD_SEND_SKB_FLAG_ACKED;
+            if (skb->flags & UCT_UD_SEND_SKB_FLAG_COMP) {
+                uct_ud_comp_desc(skb)->status = status;
+            }
+        }
+    }
+}
+
+static void uct_ud_ep_window_purge(uct_ud_ep_t *ep, ucs_status_t status)
+{
+    uct_ud_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_ud_iface_t);
+
+    uct_ud_ep_tx_stop(ep);
+    uct_ud_ep_window_release(iface, ep, ep->tx.psn, status, 0);
+    ucs_assert(ucs_queue_is_empty(&ep->tx.window));
+}
+
+static unsigned uct_ud_ep_deferred_timeout_handler(void *arg)
+{
+    uct_ud_ep_t *ep       = arg;
+    uct_ud_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_ud_iface_t);
+    ucs_status_t status;
+
+    if (ep->flags & UCT_UD_EP_FLAG_DISCONNECTED) {
+        ucs_assert(ucs_queue_is_empty(&ep->tx.window));
+        return 0;
+    }
+
+    if (ep->flags & UCT_UD_EP_FLAG_PRIVATE) {
+        ucs_assert(ucs_queue_is_empty(&ep->tx.window));
+        uct_ep_destroy(&ep->super.super);
+        return 0;
+    }
+
+    uct_ud_ep_window_purge(ep, UCS_ERR_ENDPOINT_TIMEOUT);
+
+    status = iface->super.ops->set_ep_failed(&iface->super, &ep->super.super,
+                                             UCS_ERR_ENDPOINT_TIMEOUT);
+    if (status != UCS_OK) {
+        ucs_fatal("UD endpoint %p: unhandled timeout error", ep);
+    }
+
+    return 0;
+}
+
+static void uct_ud_ep_timer_backoff(uct_ud_ep_t *ep)
+{
+    uct_ud_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_ud_iface_t);
+
+    ep->tx.tick *= iface->tx.timer_backoff;
+    ep->tx.tick  = ucs_min(ep->tx.tick, UCT_UD_SLOW_TIMER_MAX_TICK(iface));
```

Comment:
[q] maybe (`ucs_min` implementation will not compute `*` twice)
```
ep->tx.tick = ucs_min(ep->tx.tick * iface->tx.timer_backoff,
                      UCT_UD_SLOW_TIMER_MAX_TICK(iface));
```

Reply:
yes, can unite, it will not compute twice